### PR TITLE
DRILL-6446: Support for EMIT outcome in TopN

### PR DESCRIPTION
**Pull request by sohami**

### CODE REVIEW

**Comment by sohami** [comment]
@parthchandra - Please help to review.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/TopN/TopNBatch.java [hunk 1]
```diff
@@ -162,56 +173,67 @@ public void buildSchema() throws SchemaChangeException {
...
+      return getFinalOutcome();
     }
 
     try{
       outer: while (true) {
         Stopwatch watch = Stopwatch.createStarted();
-        IterOutcome upstream;
         if (first) {
-          upstream = IterOutcome.OK_NEW_SCHEMA;
+          laskKnownOutcome = IterOutcome.OK_NEW_SCHEMA;
```
Comment (parthchandra) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/TopN/TopNBatch.java]: Thought this was a typo in one place until I realised this is how it is spelt all over. :)

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/svremover/AbstractSV4Copier.java [hunk 2]
```diff
@@ -22,10 +22,12 @@
 import org.apache.drill.exec.record.VectorContainer;
 import org.apache.drill.exec.record.VectorWrapper;
 import org.apache.drill.exec.record.selection.SelectionVector4;
-import org.apache.drill.exec.vector.ValueVector;
 
 public abstract class AbstractSV4Copier extends AbstractCopier {
-  protected ValueVector[][] vvIn;
+  // Storing VectorWrapper reference instead of ValueVector[]. With EMIT outcome support underlying operator
```
Comment (parthchandra) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/svremover/AbstractSV4Copier.java]: Does this change affect RemovingRecordBatch?

File: exec/java-exec/src/main/java/org/apache/drill/exec/record/selection/SelectionVector4.java [hunk 3]
```diff
@@ -119,15 +119,38 @@ public boolean next() {
     return true;
   }
 
+  public boolean isEmpty() {
```
Comment (parthchandra) [inline exec/java-exec/src/main/java/org/apache/drill/exec/record/selection/SelectionVector4.java]: I would expect any method named isXYZ() to return a boolean without modifying the internal state.  What was wrong with leaving it the way it is?

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/TopN/TopNBatch.java [hunk 4]
```diff
@@ -162,56 +173,67 @@ public void buildSchema() throws SchemaChangeException {
...
         case STOP:
-          return upstream;
+          return laskKnownOutcome;
         case OK_NEW_SCHEMA:
           // only change in the case that the schema truly changes.  Artificial schema changes are ignored.
+          // schema change handling in case when EMIT is also seen is same as without EMIT. i.e. only if union type
+          // is enabled it will be handled.
+          container.clear();
+          firstBatchForSchema = true;
           if (!incoming.getSchema().equals(schema)) {
```
Comment (parthchandra) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/TopN/TopNBatch.java]: equals() or isEquivalent() ?

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/svremover/AbstractSV4Copier.java [hunk 5]
```diff
@@ -22,10 +22,12 @@
...
+  // operator can generate multiple output batches with no schema changes which will change the ValueVector[]
+  // reference but not VectorWrapper reference.
+  protected VectorWrapper<?>[] vvIn;
```
Comment (parthchandra) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/svremover/AbstractSV4Copier.java]: Does this require a change to RemovingRecordBatch()?

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/TopN/TopNBatch.java (hunk 1 above)
Comment (sohami) [reply exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/TopN/TopNBatch.java]: Fixed.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/svremover/AbstractSV4Copier.java (hunk 2 above)
Comment (sohami) [reply exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/svremover/AbstractSV4Copier.java]: RemovingRecordBatch uses the copier so all these changes are contained in copier only. Other than this it should not affect the behavior of RemovingRecordBatch in any other way.

File: exec/java-exec/src/main/java/org/apache/drill/exec/record/selection/SelectionVector4.java (hunk 3 above)
Comment (sohami) [reply exec/java-exec/src/main/java/org/apache/drill/exec/record/selection/SelectionVector4.java]: Reverted the change. Earlier did it to use isEmpty() method but it was not needed later on.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/TopN/TopNBatch.java (hunk 4 above)
Comment (sohami) [reply exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/TopN/TopNBatch.java]: I am keeping it as is for now. Don't want to change any behavior related to schema change handling.

**Comment by sohami** [comment]
Rebased on top of master and also handled issue with the case when first input batch with EMIT outcome is empty.

**Pull request closed without merging**